crates/astria-sequencer/src/accounts/state_ext.rs

```diff
@@ -65,59 +69,131 @@ fn nonce_storage_key<T: AddressBytes>(address: T) -> String {
+        &self,
+        address: impl AddressBytes,
+    ) -> AccountAssetsStream<Self::PrefixKeysStream> {
         let prefix = format!("{}/balance/", StorageKey(&address));
-        let mut balances: Vec<AssetBalance> = Vec::new();
-
-        let mut stream = std::pin::pin!(self.prefix_keys(&prefix));
-        while let Some(Ok(key)) = stream.next().await {
-            let Some(value) = self
-                .get_raw(&key)
-                .await
-                .context("failed reading raw account balance from state")?
-            else {
-                // we shouldn't receive a key in the stream with no value,
-                // so this shouldn't happen
-                continue;
-            };
-
-            let asset = key
-                .strip_prefix(&prefix)
-                .context("failed to strip prefix from account balance key")?
-                .parse::<crate::storage_keys::hunks::Asset>()
-                .context("failed to parse storage key suffix as address hunk")?
-                .get();
-
-            let Balance(balance) =
-                Balance::try_from_slice(&value).context("invalid balance bytes")?;
-
-            let native_asset = self
-                .get_native_asset()
-                .await
-                .context("failed to read native asset from state")?;
-            if asset == native_asset.to_ibc_prefixed() {
-                balances.push(AssetBalance {
-                    denom: native_asset.into(),
-                    balance,
-                });
-                continue;
-            }
+        AccountAssetsStream {
+            underlying: self.prefix_keys(&prefix),
+        }
+    }
 
-            let denom = self
-                .map_ibc_to_trace_prefixed_asset(asset)
-                .await
-                .context("failed to get ibc asset denom")?
-                .context("asset denom not found when user has balance of it; this is a bug")?
-                .into();
-            balances.push(AssetBalance {
-                denom,
-                balance,
-            });
+    #[instrument(skip_all)]
+    fn account_asset_balances(
+        &self,
+        address: impl AddressBytes,
+    ) -> AccountAssetBalancesStream<Self::PrefixRawStream> {
+        let prefix = format!("{}/balance/", StorageKey(&address));
+        AccountAssetBalancesStream {
+            underlying: self.prefix_raw(&prefix),
         }
-        Ok(balances)
+    }
+
+    #[instrument(skip_all)]
+    async fn get_account_balances(
+        &self,
+        address: impl AddressBytes,
+    ) -> anyhow::Result<std::collections::HashMap<asset::IbcPrefixed, u128>> {
+        use futures::TryStreamExt as _;
+        self.account_asset_balances(address)
+            .map_ok(
+                |crate::accounts::AssetBalance {
+                     asset,
+                     balance,
+                 }| (asset, balance),
+            )
+            .try_collect::<std::collections::HashMap<_, _>>()
```

Comment:
Add a note there that this relies on the IBC prefixed assets coming out of the stream being unique.

This assumption is justified because the prefixes are extracted from (unique) keys (I certainly hope they are unique because otherwise that would be a weird database).